### PR TITLE
Improve performance

### PR DESCRIPTION
**Pull request by yuliaveres**

### CODE REVIEW

File: Token/APIClients/ChatAPIClient.swift [hunk 1]
```diff
@@ -34,18 +34,24 @@ public class ChatAPIClient: NSObject {
...
-                print(response)
-                print(json ?? "")
+        DispatchQueue.global(qos: .userInitiated).async {
+            
+            self.teapot.get("/v1/accounts/bootstrap/") { (result: NetworkResult) in
+                
+                DispatchQueue.main.async {
+                    switch result {
+                    case .success(let json, let response):
+                        guard response.statusCode == 200 else { fatalError("Could not retrieve timestamp from chat service.") }
```
Comment (Elland) [inline Token/APIClients/ChatAPIClient.swift]: I think we can move timestamp server from crashing the app to returning an error to the user, asking them to try again? And even adding a retry call ourselves first?

File: Token/APIClients/ChatAPIClient.swift [hunk 2]
```diff
@@ -60,25 +66,31 @@ public class ChatAPIClient: NSObject {
             let hashedPayload = cereal.sha3WithID(string: payloadString)
             let message = "PUT\n\(path)\n\(timestamp)\n\(hashedPayload)"
             let signature = "0x\(cereal.signWithID(message: message))"
-
+
```
Comment (Elland) [inline Token/APIClients/ChatAPIClient.swift]: Lint! 👍

File: Token/APIClients/ChatAPIClient.swift [hunk 3]
```diff
@@ -60,25 +66,31 @@ public class ChatAPIClient: NSObject {
...
+            
+            DispatchQueue.global(qos: .userInitiated).async {
+                
+                self.teapot.put(path, parameters: requestParameter, headerFields: fields) { result in
+                    
+                    DispatchQueue.main.async {
+                        switch result {
+                        case .success(_, let response):
+                            guard response.statusCode == 204 else {
+                                print("Could not register user. Status code \(response.statusCode)")
```
Comment (Elland) [inline Token/APIClients/ChatAPIClient.swift]: Would be great if we could get this reported back to us, too!

File: Token/Models/ContactsManager.h [hunk 4]
```diff
@@ -28,14 +28,13 @@
 
 @interface ContactsManager : NSObject <ContactsManagerProtocol>
 
+@property (nonatomic, copy, readonly, nonnull) NSArray<TokenUser *> *tokenContacts;
```
Comment (Elland) [inline Token/Models/ContactsManager.h]: And protocol compliance worked? Nice!

File: Token/AvatarManager.swift [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import Foundation
+import Teapot
+import AwesomeCache
+
+final class AvatarManager: NSObject, CacheExpiryDefault {
```
Comment (Elland) [inline Token/AvatarManager.swift]: 😍

File: Token/Models/ContactsManager.h (hunk 4 above)
Comment (yuliaveres) [inline Token/Models/ContactsManager.h]: why might not ?

Comment (Elland) [inline Token/Models/ContactsManager.h]: Property v. method. I thought it wouldn't but ObjctiveC has it's quirks I guess. Also apparently I commented on the wrong line, supposed to be the one below, for the signal stuff.

File: Token/APIClients/ChatAPIClient.swift (hunk 3 above)
Comment (faymontage) [inline Token/APIClients/ChatAPIClient.swift]: (yes, but in a different PR)

Comment (Elland) [inline Token/APIClients/ChatAPIClient.swift]: @faymontage we can do something like that? ✨nice!

File: Token/APIClients/ChatAPIClient.swift (hunk 1 above)
Comment (yuliaveres) [inline Token/APIClients/ChatAPIClient.swift]: i think that is also not related

Comment (Elland) [inline Token/APIClients/ChatAPIClient.swift]: Should I add a Trello card?

Comment (yuliaveres) [inline Token/APIClients/ChatAPIClient.swift]: yes, backlog

File: Token/APIClients/AppsAPIClient.swift [hunk 6]
```diff
@@ -30,43 +30,56 @@ public class AppsAPIClient: NSObject, CacheExpiryDefault {
...
-                    return
-                }
-
-                let appsJSON = json["results"] as! [[String: Any]]
-                let apps = appsJSON.map { json -> TokenUser in
-                    let app = TokenUser(json: json)
-
-                    return app
+        
+        DispatchQueue.global(qos: .userInitiated).async {
```
Comment (yuliaveres) [inline Token/APIClients/AppsAPIClient.swift]: we may try this

Comment (yuliaveres) [inline Token/APIClients/AppsAPIClient.swift]: but is that actually needed ?

File: Token/APIClients/AppsAPIClient.swift [hunk 7]
```diff
@@ -30,43 +30,56 @@ public class AppsAPIClient: NSObject, CacheExpiryDefault {
...
+            self.teapot.get("/v1/apps/featured?limit=32") { (result: NetworkResult) in
+                
+                DispatchQueue.main.async {
```
Comment (yuliaveres) [inline Token/APIClients/AppsAPIClient.swift]: oh, in that case no, sure

Comment (yuliaveres) [inline Token/APIClients/AppsAPIClient.swift]: thanks Teapot documentation!
`- completion: The completion block, called with a NetworkResult once the request completes.`

File: Token/APIClients/AppsAPIClient.swift (hunk 6 above)
Comment (Elland) [inline Token/APIClients/AppsAPIClient.swift]: I think it results in cleaner calls, not sure if there's any measurable performance improvement to be had tho. More of an organisation thing.

File: Token/APIClients/AppsAPIClient.swift (hunk 7 above)
Comment (Elland) [inline Token/APIClients/AppsAPIClient.swift]: Fixing it there now as well.

File: Token/Models/ContactsManager.m [hunk 8]
```diff
@@ -119,44 +118,40 @@ - (nullable TokenUser *)tokenContactForAddress:(nullable NSString *)address
 
 - (NSArray<SignalAccount *> *)signalAccounts
 {
-    if (!_signalAccounts) {
```
Comment (Elland) [inline Token/Models/ContactsManager.m]: I missed this. You shouldn't cache it! 👌🏽

**Comment by Elland** [comment]
No assignee designated here, btw.
